Clear traverser before traversal operations in dfu

milroy opened the pull request:
Recent debugging revealed that the traverser error message wasn't getting cleared after some traversal operations, leading to spurious or error logging that can be unrelated to the error cause.

Wrap each traverser function call with `clear_err_message ()` to ensure error messages are correctly attributed to the cause.

jameshcorbett reviewed (approved):
Looks reasonable and harmless. Did you verify this makes error messages more reasonable?

milroy commented:
> Did you verify this makes error messages more reasonable?

I'm looking at the test cases to see if there's a straightforward way to check that the message has been cleared.

trws reviewed (changes requested):
I like what this is doing, but I'd prefer to have the `clear_err_message` calls at the beginning of the mark, run and find functions rather than their callers so we get that consistently between the module and command callers. If I'm reading it right, it should work to put them at the start of the methods in the dfu_traverser_t type in dfu.cpp.

milroy commented:
> If I'm reading it right, it should work to put them at the start of the methods in the dfu_traverser_t type in dfu.cpp.

That makes sense. I'll update the commit shortly.

milroy commented:
@trws, I updated the PR to address your change request and added an additional commit to remove the error message clearing from the resource match callers.

Let me know if the PR needs additional changes.

milroy commented:
I just remembered that the error message is actually member data of `dfu_impl_t` and `dfu_traverser_t` calls methods that change the error message.

I think we want to clear the error message at the beginning of methods in `dfu_impl_t`.

trws reviewed (approved):
Looks great, thanks @milroy!

trws commented:
> I just remembered that the error message is actually member data of `dfu_impl_t` and `dfu_traverser_t` calls methods that change the error message.
> 
> I think we want to clear the error message at the beginning of methods in `dfu_impl_t`.

I see what you mean, but I _think_ it's already where we want it.  The error message accumulates errors or issues as it passes through `dfu_impl_t` methods (hence all the `+= err` stuff).  With the clears in the `dfu_traverser_t` we should get the full error from whatever `dfu_impl_t` operations it ran as part of the traverser operation, where if we clear at the start of each of the `impl` methods we'd only get the last one unless we ensure that we propagate any error eagerly.

milroy commented:
> With the clears in the dfu_traverser_t we should get the full error from whatever dfu_impl_t operations it ran as part of the traverser operation, where if we clear at the start of each of the impl methods we'd only get the last one unless we ensure that we propagate any error eagerly.

That's a good point. I'll leave the clear calls in the `dfu_traverser_t` public API.

milroy commented:
The PR is good to go. I'm setting MWP. Thanks for the feedback @trws and @jameshcorbett!

milroy commented:
> I'm looking at the test cases to see if there's a straightforward way to check that the message has been cleared.

As an update, I can't think of an easy way to add a check for an uncleared error message.
